components/eam/src/control/sat_hist.F90

```diff
@@ -501,92 +521,141 @@ subroutine sat_hist_write( tape , nflds, nfils)
 
     type(File_desc_t),intent(inout)  :: File
-    type(hentry),     intent(in), target     :: hitem
+    type(hentry),     intent(in), target     :: hitems(:)
+    integer,          intent(in)     :: nflds
     integer,          intent(in)     :: ncols
+    integer,          intent(in)     :: nlevs
     integer,          intent(in)     :: nfils
     integer,          intent(in)     :: fdims(:)
     integer,          intent(in)     :: ldims(:)
     integer,          intent(in)     :: owners(:)
+    integer,          intent(in)     :: decomp
+
 
     type(field_info), pointer :: field
     type(var_desc_t) :: vardesc
-    type(iosystem_desc_t), pointer :: sat_iosystem
-    type(io_desc_t) :: iodesc
-    integer :: t, ierr, ndims
-    integer, allocatable :: dimlens(:)
+    integer :: ierr
 
-    real(r8), allocatable :: buf(:)
-    integer,  allocatable :: dof(:)
-    integer :: i,k, cnt
+    real(r8) :: sbuf1d(ncols),rbuf1d(ncols)
+    real(r4) :: buf1d(ncols)
+    real(r8) :: sbuf2d(nlevs,ncols), rbuf2d(nlevs,ncols)
+    real(r4) :: buf2d(nlevs,ncols)
+    integer :: i,k,f, cnt
 
     call t_startf ('sat_hist::dump_columns')
 
-    sat_iosystem => File%iosystem
-    field => hitem%field
-    vardesc = hitem%varid(1)
-
-
-    ndims=1
-    if(associated(field%mdims)) then
-       ndims = size(field%mdims)+1
-    else if(field%numlev>1) then
-       ndims=2
-    end if
-    allocate(dimlens(ndims))
-    dimlens(ndims)=ncols
-    if(ndims>2) then
-       do i=1,ndims-1
-          dimlens(i)=hist_coords(field%mdims(i))%dimsize
-       enddo
-    else if(field%numlev>1) then
-       dimlens(1) = field%numlev
-    end if
-   
-
-    allocate( buf( product(dimlens) ) )
-    allocate( dof( product(dimlens) ) )
+    do f = 1,nflds
+       field => hitems(f)%field
+
+       if (field%numlev==nlevs .and. field%decomp_type==decomp) then
+          vardesc = hitems(f)%varid(1)
+
+          if (nlevs==1) then
+             sbuf1d = 0.0_r8
+             rbuf1d = 0.0_r8
+             do i=1,ncols
+                if ( iam == owners(i) ) then
+                   sbuf1d(i) = hitems(f)%hbuf( fdims(i), 1, ldims(i) )
+                endif
+             enddo
+             call mpi_allreduce(sbuf1d,rbuf1d,ncols,mpi_real8, mpi_sum, mpicom, ierr)
+             buf1d(:) = real(rbuf1d(:),r4)
+             ierr = pio_put_var(File, vardesc, (/nfils/),(/ncols/), buf1d(:))
+          else
+             sbuf2d = 0.0_r8
+             rbuf2d = 0.0_r8
+             do i=1,ncols
+                if ( iam == owners(i) ) then
+                   do k = 1,nlevs
+                      sbuf2d(k,i) = hitems(f)%hbuf( fdims(i), k, ldims(i) )
+                   enddo
+                endif
+             enddo
+             call mpi_allreduce(sbuf2d,rbuf2d,ncols*nlevs,mpi_real8, mpi_sum, mpicom, ierr)
+             buf2d(:,:) = real(rbuf2d(:,:),r4)
+             ierr = pio_put_var(File, vardesc, (/1,nfils/),(/nlevs,ncols/), buf2d(:,:))
```

Comment:
(ok, looking at the code again, the mpi_allreduce() seems to be just a way to collect the column data across processes, so maybe you are not really ending up with the sum across columns - just adding 0s from non-owners of columns. But I think the new version is doing more work - collecting column data at all procs and using put_var to write the data. The older version created a dof based on local data and used it to write the column data in parallel)

Reply:
If its not a performance critical part of the code (only enabled for debugging etc) your current change should be ok for now since the old code wasn't working anyway. You might want to add a FIXME in the code or have an issue as a reminder to look at it again.

Reply:
a FIXME is fine.